src/levanter/models/whisper.py

```diff
@@ -0,0 +1,562 @@
+
+        return WhisperLayer(self_attn, attn_ln, encoder_attn, encoder_attn_ln, mlp, mlp_ln)
+
+    @named_call
+    def __call__(
+        self, x: NamedArray, xa: Optional[NamedArray] = None, mask: Optional[AttentionMask | NamedArray] = None, *, key
+    ):
+        k1, k2, k3 = haliax.jax_utils.maybe_rng_split(key, 3)
+
+        attn_output = self.self_attn(self.attn_ln(x), mask=mask, key=k1)
+        x = x + attn_output
+
+        if self.encoder_attn:
+            xs = x
+            if self.encoder_attn_ln:
+                xs = self.encoder_attn_ln(xs)
+            x = x + self.encoder_attn(xs, xa, key=k2)
+
+        ff_output = self.mlp(self.mlp_ln(x), key=k3)
+        x = x + ff_output
+
+        return x
+
+    def _state_dict_key_map(self) -> Dict[str, Optional[str]]:
+        return {
+            "mlp": None,
+            "mlp_ln": "final_layer_norm",
+            "attn_ln": "self_attn_layer_norm",
+            "encoder_attn_ln": "encoder_attn_layer_norm",
+        }
+
+
+class WhisperTransformer(eqx.Module, StateDictSerializationMixin):
+    layers: Stacked[WhisperLayer]
+    Layer: Axis
+    ln_f: hnn.LayerNorm
+
+    @staticmethod
+    def init(Layer: Axis, Heads: Axis, HeadSize: Axis, config: WhisperConfig, has_cross: bool, *, key):
+        # vectorize the blocks
+        layers = Stacked.init(Layer, WhisperLayer, gradient_checkpointing=config.gradient_checkpointing)(
+            Heads,
+            HeadSize,
+            config,
+            has_cross=has_cross,
+            key=shaped_rng_split(key, Layer.size),
+        )
+        ln_f = hnn.LayerNorm.init(config.Embed, eps=config.layer_norm_epsilon, use_bias=config.use_bias)
+
+        return WhisperTransformer(layers, Layer, ln_f)
+
+    @named_call
+    def __call__(
+        self,
+        x: NamedArray,
+        xa: Optional[NamedArray] = None,
+        attn_mask: Optional[AttentionMask | NamedArray] = None,
+        *,
+        key=None,
+    ) -> NamedArray:
+        keys = hax.jax_utils.maybe_rng_split(key, self.Layer) if key is not None else None
+        x = self.layers.fold(x, xa, attn_mask, key=keys)
+        x = self.ln_f(x)
+
+        return x
+
+    def from_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None):
+        stacked = stack_state_dict(state_dict, prefix=apply_prefix(prefix, "layers"))
+        out = super().from_state_dict(stacked, prefix=prefix)
+        return out
+
+    def update_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None) -> StateDict:
+        my_state_dict: StateDict = {}
+        super().update_state_dict(my_state_dict, prefix=prefix)
+
+        stacked_dict = unstack_state_dict(my_state_dict, prefix=apply_prefix(prefix, "layers"))
+        state_dict.update(stacked_dict)
+
+        return state_dict
+
+    def _state_dict_key_map(self) -> Dict[str, Optional[str]]:
+        return {"ln_f": "layer_norm"}
```

Comment:
you can also just call your member variable layer_norm and not need this, entirely your call

Reply:
Shifted it here! In general, I like the shorter form names but it does seem silly to have the full map function just for the tiny name difference.